### ISSUE TITLE
<Tab> auto-complete for npm prints out junk

### ISSUE THREAD

**Issue by nomaed** [fish-shell/fish-shell#3152]
In the last day or two I have been getting junk output when I am trying to get auto-complete for `npm`.
On another computer it works fine, and I don't seem to have any problems with other commands.

I tried rebuilding the completions, and it didn't help.
I don't see anything specific errors or function names that I can understand what's going on since I'm pretty new to Fish.

After the garbage, there are normal commands but some garbage also appears in between.
If anyone can help with suggestion where to look, it'd be great.

---
Operating system: OS X El Capitan 10.11.5 (15F34)
Fish version: 2.3.0
Fish installed: using the official DMG.
Fish location: /usr/local/bin/fish
Terminal: iTerm2

---

Output:
```
$ npm -v
\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ 2\>/dev/null\)\)\ \|\|\ return\ \$\?
\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ COMP_LINE=\"\$COMP_LINE\"\ \\
\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ COMP_POINT=\"\$COMP_POINT\"\ \\
\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ npm\ completion\ --\ \"\$\{words\[@\]\}\"\ \\
\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ 2\>/dev/null\)\)\ \|\|\ return\ \$\?
\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ COMP_LINE=\"\$line\"\ \\
\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ COMP_POINT=\"\$point\"\ \\
\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ npm\ completion\ --\ \"\$\{words\[@\]\}\"\ \\
\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ 2\>/dev/null\)
\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ COMP_LINE=\$BUFFER\ \\
\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ COMP_POINT=0\ \\
\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ npm\ completion\ --\ \"\$\{words\[@\]\}\"\ \\
\ \ \ \ \ \ _get_comp_words_by_ref\ -n\ =\ -n\ @\ -w\ words\ -i\ cword
\ \ \ \ \ \ cword=\"\$COMP_CWORD\"
\ \ \ \ \ \ words=\(\"\$\{COMP_WORDS\[@\]\}\"\)
\ \ \ \ compadd\ --\ \$\(COMP_CWORD=\$\(\(CURRENT-1\)\)\ \\
\ \ \ \ else
\ \ \ \ fi
\ \ \ \ if\ type\ _get_comp_words_by_ref\ \&\>/dev/null\;\ then
\ \ \ \ IFS=\"\$si\"
\ \ \ \ IFS=\$\'\\n\'\ COMPREPLY=\(\$\(COMP_CWORD=\"\$cword\"\ \\
\ \ \ \ IFS=\$\'\\n\'\ reply=\(\$\(COMP_CWORD=\"\$cword\"\ \\
\ \ \ \ IFS=\$si
\ \ \ \ let\ cword-=1
\ \ \ \ local\ cword\ line\ point\ words\ si
\ \ \ \ local\ si=\"\$IFS\"
\ \ \ \ local\ si=\$IFS
\ \ \ \ local\ words\ cword
\ \ \ \ read\ -Ac\ words
\ \ \ \ read\ -cn\ cword
\ \ \ \ read\ -l\ line
\ \ \ \ read\ -ln\ point
\ \ \ \ si=\"\$IFS\"
\ \ _npm_completion\ \(\)\ \{
\ \ _npm_completion\(\)\ \{
\ \ compctl\ -K\ _npm_completion\ npm
\ \ compdef\ _npm_completion\ npm
\ \ complete\ -o\ default\ -F\ _npm_completion\ npm
\ \ \}
\#
\#\ Installation:\ npm\ completion\ \>\>\ \~/.bashrc\ \ \(or\ \~/.zshrc\)
\#\ npm\ command\ completion\ script
\#\ Or,\ maybe:\ npm\ completion\ \>\ /usr/local/etc/bash_completion.d/npm
\#\#\#-begin-npm-completion-\#\#\#
\#\#\#-end-npm-completion-\#\#\#
add-user                                                              (Add a registry user account)
adduser                                                               (Add a registry user account)
bin                                                                        (Display npm bin folder)
bugs                                                    (Bugs for a package in a web browser maybe)
c                                                              (Manage the npm configuration files)
cache                                                                 (Manipulates package's cache)
config                                                         (Manage the npm configuration files)
ddp                                                                            (Reduce duplication)
dedupe                                                                         (Reduce duplication)
deprecate                                                        (Deprecate a version of a package)
docs                                                    (Docs for a package in a web browser maybe)
edit                                                                    (Edit an installed package)
elif\ type\ compctl\ \&\>/dev/null\;\ then
elif\ type\ compdef\ \&\>/dev/null\;\ then
explore                                                               (Browse an installed package)
faq                                                                    (Frequently Asked Questions)
fi
find                                                                          (Search for packages)
find-dupes                                                                     (Reduce duplication)
get                                                               (Echo the config value to stdout)
help-search                                                         (Search npm help documentation)
home                                                    (Docs for a package in a web browser maybe)
i                                                                               (install a package)
if\ type\ complete\ \&\>/dev/null\;\ then
info                                                                           (View registry info)
install                                                                         (install a package)
isntall                                                                         (install a package)
issues                                                  (Bugs for a package in a web browser maybe)
la                                                                        (List installed packages)
link                                                                     (Symlink a package folder)
rows 1 to 75 of 118
```

**Comment by krader1961** [comment]
The first block of code looks like the script that running `npm completion` would create for bash or zsh. Fish also runs that command but manipulates the output before using it so that it is compatible with fish syntax. Did you by any chance upgrade to fish 2.3.0 a couple of days ago? If yes did you also restart all your existing shell instances? In a shell session that exhibits the problem what does `echo $FISH_VERSION` output?

**Comment by nomaed** [comment]
I only installed Fish several days ago, and I didn't do any re-installation.
$FISH_VERSION is 2.3.0
And I am sure that it used to work. The only thing of any significance that I did was rebuilding neovim and remove and reinstalling python using homebrew, but I doubt it should have anything to do with it.

**Comment by krader1961** [comment]
I don't see how reinstalling python could be relevant. Are you running fish inside neovim when the problem occurs? If yes does it happen in a plain iTerm2 fish session? Is the system no exhibiting the problem also running fish 2.3.0?

Does the problem disappear if you rename */usr/local/share/fish/completions/npm.fish* and start a new shell? Obviously that should cause no completions other than the core fish pathname completions to occur but let's confirm that fish is using the completion script it should be using.

**Comment by nomaed** [comment]
Nope, not running in neovim, just in iTerm2. 
I'm getting the same in the default terminal as well.
If I rename npm.fish to something else, I do get the standard path completions. Even before starting a new shell.

I'm still trying to figure what's the difference between my fish on the OS X and the Ubuntu installation on which it works fine. If I'll notice any diff, I'll update here.

**Comment by krader1961** [comment]
Lots of things could be a factor. Obviously start by ensuring the *npm.fish* script on the problem machine is identical to the copy on the non-problem machine. Then I'd check that `locale` is the same on both machines. Verify that the npm command is the same version on both machines and if it has any configurable options that they're the same. I'm not a NodeJS user so I can't offer more specific advice but given that only fish's npm completions are weird it seems highly likely there is a difference in the NodeJS environment.

**Comment by faho** [comment]
@nomaed: The npm completions call `npm completion` (the bash/zsh completion tool that AFAIK ships with npm). It's possible you don't have that setup correctly.

The messages you get as completion options (which are what that tool prints to stdout) look to me to be instructions to set it up for bash/zsh (only sorted). Of course it's also possible the tool changed.

**Comment by nomaed** [comment]
I looked at the histories of both NPM's completion code and npm.fish and I see that nothing changed for a while. Also, my npm.fish has no diff from the one in the repo.
I tried switching to various npm versions, including to what I have on the Linux, and I tried different node versions via nvm (also tried with system version). No difference, I am getting the same output.

Locale hasn't changed either.

I guess I'll have to debug npm.fish line by line :)
